utility_network/perform-valve-isolation-trace/src/main/java/com/esri/samples/perform_valve_isolation_trace/PerformValveIsolationTraceController.java

```diff
@@ -240,64 +237,59 @@ private void handleTraceClick() {
-            var utilityElementTraceResult = (UtilityElementTraceResult) utilityTraceResults.get(0);
+      utilityNetwork.traceAsync(utilityTraceParameters).toCompletableFuture().whenComplete(
+        (utilityTraceResults, ex) -> {
+          if (ex == null) {
+            if (utilityTraceResults.get(0) instanceof UtilityElementTraceResult) {
+              var utilityElementTraceResult = (UtilityElementTraceResult) utilityTraceResults.get(0);
 
-            if (!utilityElementTraceResult.getElements().isEmpty()) {
+              if (!utilityElementTraceResult.getElements().isEmpty()) {
 
-              // iterate through the map's feature layers
-              mapView.getMap().getOperationalLayers().forEach(layer -> {
-                if (layer instanceof FeatureLayer) {
+                // iterate through the map's feature layers
+                mapView.getMap().getOperationalLayers().forEach(layer -> {
+                  if (layer instanceof FeatureLayer) {
 
-                  // create query parameters to find features whose network source name matches the layer's feature
-                  // table name
-                  var queryParameters = new QueryParameters();
-                  utilityElementTraceResult.getElements().forEach(utilityElement -> {
+                    // create query parameters to find features whose network source name matches the layer's feature
+                    // table name
+                    var queryParameters = new QueryParameters();
+                    utilityElementTraceResult.getElements().forEach(utilityElement -> {
 
-                    String networkSourceName = utilityElement.getNetworkSource().getName();
-                    String featureTableName = ((FeatureLayer) layer).getFeatureTable().getTableName();
-
-                    if (networkSourceName.equals(featureTableName)) {
-                      queryParameters.getObjectIds().add(utilityElement.getObjectId());
-                    }
-                  });
+                      String networkSourceName = utilityElement.getNetworkSource().getName();
+                      String featureTableName = ((FeatureLayer) layer).getFeatureTable().getTableName();
 
-                  // select features that match the query
-                  ListenableFuture<FeatureQueryResult> featureQueryResultListenableFuture =
-                          ((FeatureLayer) layer).selectFeaturesAsync(queryParameters, FeatureLayer.SelectionMode.NEW);
+                      if (networkSourceName.equals(featureTableName)) {
+                        queryParameters.getObjectIds().add(utilityElement.getObjectId());
+                      }
+                    });
+
+                    // select features that match the query
+                    ((FeatureLayer) layer).selectFeaturesAsync(queryParameters, FeatureLayer.SelectionMode.NEW)
+                      .toCompletableFuture().thenRun(() -> {
```

Comment:
I'm not sure about this one - seems like since `selectFeaturesAsync` has a result you should check the result with a `whenComplete` or `excptionally`.

Reply:
The original `ListenableFuture` here didn't catch any exceptions either - probably because launching the sample and pressing "trace" without changing any other setting throws a `JsonEmbeddedException`. Doesn't crash the application at all, but it's not clear to me what's causing it or how to fix it (maybe that's why it was never caught in the first place?).

Reply:
Hmm. So previously, we're updating/resetting the UI regardless of what happens - which I think is right, and so `thenRun` is good to ensure we do that regardless. But I think it was a weakness before to not be checking, as if someone tinkered with the code and ended up passing null queryParameters for example, this could throw. So, dependent on @colinanderson 's feedback on this approach above... https://github.com/Esri/arcgis-maps-sdk-java-samples/pull/788#discussion_r1218193848

...maybe we can alter it to something like:
```java
                    layer.selectFeaturesAsync(queryParameters, FeatureLayer.SelectionMode.NEW)
                      .toCompletableFuture().exceptionally(exception -> {
                        new Alert(Alert.AlertType.ERROR, "Error selecting features: " + exception.getMessage());
                        return null;
                        }).thenRun(() -> {
                        // update the status text, enable the buttons and hide the progress indicator
                        statusLabel.setText("Isolation trace completed.");
                        enableUI(true);
                      });
```

Reply:
Done a bit of poking around on this: we only seem to get the exception when the combobox is set to "Subnetwork Controller", the checkbox is unselected, and the user hasn't placed any filter barriers on the map. The `JsonEmbeddedException` is fairly unhelpful: message is "Unable to complete operation." and details "Unable to perform query operation." I'm wondering if it's best to split off this problem as a new issue.

Reply:
Can also replicate using the Isolating option too. I'll create an issue as this isn't specific to completable futures. Replace with the below to better match our other implementations of CF:
```
                    // select features that match the query
                    layer.selectFeaturesAsync(queryParameters, FeatureLayer.SelectionMode.NEW)
                      .toCompletableFuture().whenComplete((result, exception) -> {
                        // update the status text, enable the buttons and hide the progress indicator
                        statusLabel.setText("Isolation trace completed.");
                        enableUI(true);
                      });
```
And we will resolve the exception handling separately since there isn't currently any either.